### DISCUSSION TITLE
mixcr align - question

### DISCUSSION THREAD

**Discussion by silvia1234567890** [milaboratory/mixcr#1404] [Q&A] [answered]
Hello,

I would like to align a fasta file with the 100 most abundant sequences against a VDJ gene segment library that I built before with the command "mixcr buildLibrary". The input fasta file contains sequences like this: 
```
>AACCAGCAAATCACC_CONSCOUNT_134
ACAGCACGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAACAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGGTGACTACTTCACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGGAGCAGATATACTGGAGCTGCATACTACAAAGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGCCTGAAGACACTGCTGTGTATTACTGTGCCAGAAACTGGGAGGGTGGATACGACTACTTTGACTACTGGGGGAAAGGGACACAAGTGACCGTGACTTCTGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGA----------------
>ATACGCTATGCAACC_CONSCOUNT_119
ACAGCACGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAACAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGGTGACTACTTCACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGGAGCAGATATACTGGAGCTGCATACTACAAAGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGCCTGAAGACACTGCTGTGTATTACTGTGCCAGAGAGTGGGAGTCCGGGTGGTTCTACTTTGACTACTGGGGAAAAGGAACAATGGTGACAGTCACAACAGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGA----------------
>GATAGCACTGGATGG_CONSCOUNT_96
ACAGCAGGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAACAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGGTGACTACTACACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGAATGAGATCTACTGGATATACATTCTACAAAGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGCCTGAAGACACTGCTGTGTATTACTGTGCCAGAAATCACTTGTGCAGCGGGTGCCCCTTTGACTACTGGGGGAAAGGCACCATGGTCACCATCACCTCGGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGA----------------
```
These are IgM sequences, and the specie is _S. maximus_... 
What parameters should I use with "mixcr align"?

Thank you in advance.
Best regards,
Silvia

**Comment by mizraelson** [comment]
Hi,
The parameters for `mixcr align` should be set to match the library structure of your sequences. If you don't have any adaptors, the data is derived from RNA (meaning no introns) and your sequences should partially cover C gene, the command would be:

```
mixcr align --preset generic-amplicon \
--library library_file_name \
--species maximus \
--rna \
--rigid-left-alignment-boundary \
--rigid-right-alignment-boundary C \
input.fasta \
output.vdjca
```
The `species` parameter should be set to the name you used when creating your library.

Sincerely,
Mark

**Reply by silvia1234567890** [reply]
Perfect. Just to be sure about the parameters I should use..

The input file I will use to do the alignment with MiXCR are pre-processed sequences. 

The raw sequences are paired-end sequencing reads with UMI, Illumina adapters, PCR primers, etc. I have pre-processed the raw sequences (removed Illumina adaptors, filtered low quality reads, UMI annotation, removal of PCR primers, generated UMI consensus sequences, paired-end assembly of UMI consensus sequences and filtering, among other things).

The pre-processed sequences are the most abundant sequences. 

To build the library I used this command:

```
mixcr buildLibrary --debug \
  --v-genes-from-fasta IGHV.fasta \
  --v-gene-feature VRegion \  
  --j-genes-from-fasta IGHJ.fasta \
  --d-genes-from-fasta IGHD.fasta \ 
  --chain IGH \
  --taxon-id 52904 \
  --species Scophthalmus_maximus \
  Scophthalmus_maximus-IGH.json.gz
```

I ran your code and I got this error message:

```
WARNING: forcing -OvParameters.geneFeatureToAlign=VRegionWithP since current gene feature (VTranscriptWithP) is absent in 100% of V genes.
Alignment: progress unknown
Please copy the following information along with the stacktrace:
   Version: 4.5.0; built=Fri Sep 22 14:39:05 CEST 2023; rev=cdb24b4fb7; lib=repseqio.v3.0.1
        OS: Mac OS X
      Java: 21
  Cmd args: align --preset generic-amplicon --library Scophthalmus_maximus-IGH.json --species Scophthalmus_maximus --rna --rigid-left-alignment-boundary --rigid-right-alignment-boundary C Turbot_SP19.fasta output.vdjca
picocli.CommandLine$ExecutionException: Error while running command align java.lang.IllegalArgumentException: Letter '-' is not defined in 'Alphabet{nucleotide}'.
	at com.milaboratory.mixcr.cli.Main.registerExceptionHandlers$lambda-12(SourceFile:340)
	at picocli.CommandLine.execute(CommandLine.java:2088)
	at com.milaboratory.mixcr.cli.Main.main(SourceFile:98)
Caused by: java.lang.IllegalArgumentException: Letter '-' is not defined in 'Alphabet{nucleotide}'.
	at com.milaboratory.core.sequence.Alphabet.parse(SourceFile:324)
	at com.milaboratory.o.ba.a(SourceFile:132)
	at com.milaboratory.o.bd.take(SourceFile:1071)
	at cc.redberry.pipe.util.SimpleProcessorWrapper.take(SimpleProcessorWrapper.java:40)
	at com.milaboratory.o.Cn.take(SourceFile)
	at cc.redberry.pipe.util.Chunk.readChunk(Chunk.java:78)
	at cc.redberry.pipe.CUtils$2.take(CUtils.java:169)
	at cc.redberry.pipe.CUtils$2.take(CUtils.java:161)
	at cc.redberry.pipe.blocks.O2ITransmitter.run(O2ITransmitter.java:63)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

what parameters should I use in my case?

Thank you in advance

**Reply by mizraelson** [reply]
The error is due to the "-" symbol in your FASTA files. Additionally, I strongly recommend using MiXCR on your raw datasets, as it efficiently handles primer trimming, UMI error correction, and quality filtering using a variety of sophisticated algorithms. The only preprocessing relevant before running MiXCR is trimming the Illumina adapters. If you can share you library structure I can help you with the command.

**Reply by silvia1234567890** [reply]
In the header of the fasta files I used to build the library I have "_" as a separator... so I will change it, thanks!
On the other hand, I would like to try the preprocessing part also with MiXCR, and then compare with which of the ways it's better for what I need... 

The raw sequences have a similar structure to the ones used in this workflow (https://presto.readthedocs.io/en/stable/workflows/VanderHeiden2017_Workflow.html) and the preprocessing steps were those. How would I have to do with MiXCR? I removed the Illumina adapters with cutadapt before the preprocessing. 

I don't think I can share the library, since the data is not yet public.. 
I hope that by knowing more or less the structure of the raw data and how the preprocessing went you can help me in the same way.

Thank you in advance.
Best regards,
Silvia

**Reply by mizraelson** [reply, marked as answer]
Its not about "_" in the library header. Its about "-" at the end of your sequences:
```
>AACCAGCAAATCACC_CONSCOUNT_134
ACAGCACGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAACAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGGTGACTACTTCACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGGAGCAGATATACTGGAGCTGCATACTACAAAGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGCCTGAAGACACTGCTGTGTATTACTGTGCCAGAAACTGGGAGGGTGGATACGACTACTTTGACTACTGGGGGAAAGGGACACAAGTGACCGTGACTTCTGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGA----------------
>ATACGCTATGCAACC_CONSCOUNT_119
ACAGCACGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAACAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGGTGACTACTTCACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGGAGCAGATATACTGGAGCTGCATACTACAAAGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGCCTGAAGACACTGCTGTGTATTACTGTGCCAGAGAGTGGGAGTCCGGGTGGTTCTACTTTGACTACTGGGGAAAAGGAACAATGGTGACAGTCACAACAGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGA----------------
>GATAGCACTGGATGG_CONSCOUNT_96
ACAGCAGGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAACAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGGTGACTACTACACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGAATGAGATCTACTGGATATACATTCTACAAAGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGCCTGAAGACACTGCTGTGTATTACTGTGCCAGAAATCACTTGTGCAGCGGGTGCCCCTTTGACTACTGGGGGAAAGGCACCATGGTCACCATCACCTCGGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGA----------------
```

Please try the following command on your raw files:

```
mixcr analyze generic-amplicon-with-umi \
    --library Scophthalmus_maximus-IGH.json \
    --species Scophthalmus_maximus \
    --rna \
    --tag-pattern "^(R1:*) \ ^(UMI:N{17})N{10}(R2:*)" \
    --rigid-left-alignment-boundary \
    --floating-right-alignment-boundary C \
      input_R1.fastq.gz \
      input_R2.fastq.gz \
      result
```
Check that in your library the UMI is also located in the first 17bp of R2 or adjust the pattern `"^(R1:*) \ ^(UMI:N{17})N{10}(R2:*)"` accordingly.

Also, if you expect reads to overlap and you are interested in assembling the full-length receptor sequence add:

`--assemble-clonotypes-by VDJRegion`

Thats it. This single command should give you the clonotype tables. Let me know how it works.

Sincerely,
Mark

**Reply by silvia1234567890** [reply]
Oh, I didn't realize, sorry. I don't have the files in front of me right now. 

Ok, I will use those commands and let you know how it works.
MiXCR is a very useful and comprehensive tool.

Thank you so much for your help!

**Reply by silvia1234567890** [reply]
Hello,

I have a question. When I run MiXCR I always get this warning:
`WARNING: forcing -OvParameters.geneFeatureToAlign=VRegionWithP since current gene feature (VTranscriptWithP) is absent in 100% of V genes.`

Like, for example:

the code I ran:

```
mixcr align --preset generic-amplicon \
> --library Scophthalmus_maximus-IGH.json \
> --species Scophthalmus_maximus \
> --rna \
> --rigid-left-alignment-boundary \
> --rigid-right-alignment-boundary C \
> Turbot_SP19.fasta \
> output.vdjca
```
what I got:

```
WARNING: forcing -OvParameters.geneFeatureToAlign=VRegionWithP since current gene feature (VTranscriptWithP) is absent in 100% of V genes.
Alignment: progress unknown
====================== report: align ======================
Analysis time: 1,14s
Total sequencing reads: 109
Successfully aligned reads: 105 (96,33%)
Coverage (percent of successfully aligned):
  CDR3: 62,86%
  FR3_TO_FR4: 100%
  CDR2_TO_FR4: 100%
  FR2_TO_FR4: 100%
  CDR1_TO_FR4: 100%
  VDJRegion: 100%
Alignment failed: absence of J hits: 4 (3,67%)
Overlapped: 0 (0%)
Overlapped and aligned: 0 (0%)
Overlapped and not aligned: 0 (0%)
Alignment-aided overlaps, percent of overlapped and aligned: 0 (NaN%)
Partial aligned reads, percent of successfully aligned: 39 (37,14%)
Realigned with forced non-floating bound: 0 (0%)
Realigned with forced non-floating right bound in left read: 0 (0%)
Realigned with forced non-floating left bound in right read: 0 (0%)
IGH chains: 105 (100%)
IGH non-functional: 0 (0%)
Trimming report:
  R1 reads trimmed left: 0 (0%)
  R1 reads trimmed right: 0 (0%)
  Average R1 nucleotides trimmed left: 0.0
  Average R1 nucleotides trimmed right: 0.0
```

Do I have to add/change/remove any parameters to fix this? Or does it not influence the result of the analysis?

Thank you

**Reply by mizraelson** [reply]
You see the warning because, by default, MiXCR expects the sequences in the reference library to contain the 5'UTR and leader sequence (`VTranscriptWithP`). In your library, the sequence starts from FR1, which is why MiXCR changes it to `VRegionWithP`. This does not affect the analysis but merely reflects the specifics of your library.
